util/src/main/java/tc/oc/pgm/util/nms/reflect/ReflectionProxy.java

```diff
@@ -0,0 +1,353 @@
+package tc.oc.pgm.util.nms.reflect;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import tc.oc.pgm.util.reflect.MinecraftReflectionUtils;
+import tc.oc.pgm.util.reflect.ReflectionUtils;
+
+public class ReflectionProxy implements InvocationHandler {
+
+  static Map<Method, Function<Object[], Object>> functionMap = new ConcurrentHashMap<>();
+
+  static ReflectionProxy INSTANCE = new ReflectionProxy();
+
+  public static <T> T getProxy(Class<T> classType) {
+    return (T)
+        Proxy.newProxyInstance(
+            ReflectionProxy.class.getClassLoader(), new Class[] {classType}, INSTANCE);
+  }
+
+  @Override
+  public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+    Class<?> declaringClass = method.getDeclaringClass();
+    Class<?> parentClass = getAnnotatedClass(declaringClass);
```

Comment:
should also go inside compute if absent, we want this to have minimal overhead after the first call